finagle/buoyant/src/main/scala/com/twitter/finagle/buoyant/IpSslClientSessionVerifier.scala

```diff
@@ -0,0 +1,83 @@
+package com.twitter.finagle.buoyant
+
+import java.net.{InetAddress, UnknownHostException}
+import java.security.cert.X509Certificate
+
+import com.twitter.finagle.{Address, SslHostVerificationException}
+import com.twitter.finagle.ssl.client.{SslClientConfiguration, SslClientSessionVerifier}
+import com.twitter.util.Try
+import javax.net.ssl.SSLSession
+import sun.security.util.HostnameChecker
+import sun.security.x509.X500Name
```

Comment:
sun.* packages are internal JDK packages and not meant to be used by applications. They are unsupported and will cause issues with upgrade to newer Java versions (9,10,etc). Can this be rewritten to remove the reliance on unsupported JDK internal classes?

Reply:
This is what finagle uses (see https://github.com/twitter/finagle/blob/finagle-18.2.0/finagle-core/src/main/scala/com/twitter/finagle/ssl/client/HostnameVerifier.scala). Two comments here
1. We could potentially copy and paste the implementation but I wouldn't get the benefit of potential fixes.
2. Apparently in the most recent version, 18.4.0 (so timely!) they removed that class. Do you mind if I create a follow up ticket to investigate alternatives? It will take some time for linkerd to adopt latest finagle anyways.

Reply:
This is the commit where they removed it: https://github.com/twitter/finagle/commit/1313d9ba77b35da0a9a6fce3599dfc7c4c5e137d

As the commit message explains, using this internal class blocks upgrade to Java 9 or higher. Starting with Java 9, Oracle has blocked applications from accessing these internal classes.

We could copy the code from the JDK: http://hg.openjdk.java.net/jdk8u/jdk8u/jdk/file/tip/src/share/classes/sun/security/util/HostnameChecker.java
But that is not the best idea, because we would be forking it and have to maintain any fixes ourselves. Also, that is GPLv2 licensed, linkerd is Apache licensed, so I'm not sure what legal issues the difference in licensing may pose.

Another option would be to implement the whole validation process ourselves. See these StackOverflow answers for pointers:
https://stackoverflow.com/a/18174689/2147204 and https://stackoverflow.com/a/17979954/2147204

Also, it may be useful to use BouncyCastle instead of the JDK for certificate parsing. It has a far richer public API; the public API of the JDK implementation is pretty limited, and often it is hard to do things without using the private API (access to which is banned in Java 9+).

OTOH, one could also just go with using the unsupported internal API for now, but mark this issue as "technical debt" to be fixed at some later point (certainly before you upgrade to Java 9/10/11).

Reply:
Let's go with the technical debt option. 
I just created https://jira.medallia.com/browse/MSF-31 to track it.
Thanks.